File: data_structure/hash_table/41_first_missing_positive.cc
#include <cmath>
#include <vector>

// 2024/01/14: 直接抄答案.
// 对于一个长度为N的数组, 其中没有出现的最小正整数只能在[1, N + 1]中.
// 使用原始数组作为哈希表.
int FirstMissingPositive(std::vector<int>& nums) {
  int n = nums.size();
  // 1. 把不在[1, n]的所有的数都修改成大于n的数, 比如n+1.
  for (int i = 0; i < n; ++i) {
    if (nums[i] < 1 || nums[i] > n) {
      nums[i] = n + 1;
    }
  }

  // 2.1 遍历数组中的每个数x, 由于其可能被打上了标记, 所以取绝对值.
  // 2.2 判断x是否在[1, n]内, 如果在, 给相应的位置打上标记.
  for (int i = 0; i < n; ++i) {
    int x = std::abs(nums[i]);
    if (x >= 1 && x <= n) {
      if (nums[x - 1] > 0) {
        nums[x - 1] = -nums[x - 1];
      }
    }
  }

  // 3. 如果每一个数都是负数, 那么答案是n+1, 否则是i+1.
  int res = n + 1;
  for (int i = 0; i < n; ++i) {
    if (nums[i] > 0) {
      return i + 1;
    }
  }
  return res;
}

// 2024/01/14: 直接抄答案, 采用置换的方法.(本质上也是标记)
int FirstMissingPositive2(std::vector<int>& nums) {
  int n = nums.size();
  for (int i = 0; i < n; ++i) {
    while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] != nums[i]) {
      std::swap(nums[nums[i] - 1], nums[i]);
    }
  }
  for (int i = 0; i < n; ++i) {
    if (nums[i] != i + 1) {
      return i + 1;
    }
  }
  return n + 1;
}

File: data_structure/hash_table/41_first_missing_positive.py
from typing import List


# 2026/05/11: 自己写出答案, 但是这种方式空间复杂度为 O(n), 不满足需求.
def first_missing_positive(nums: List[int]) -> int:
    nums_set = set(nums)
    nums_set.add(0)
    res_list = []
    for num in nums_set:
        if num >= 0 and num + 1 not in nums_set:
            res_list.append(num + 1)
    return min(res_list)


# 2026/05/11: 直接抄答案.
# 1. 对于一个长度为 N 的数组, 其中没有出现的最小正整数只能在[1, N + 1]中.
# 2. 使用原始数组作为哈希表以实现空间复杂度为常数级别.
def first_missing_positive2(nums: List[int]) -> int:
    length = len(nums)
    for i in range(length):
        if nums[i] < 1 or nums[i] > length:
            nums[i] = length + 1

    for num in nums:
        abs_num = abs(num)
        if abs_num >= 1 and abs_num <= length:
            nums[abs_num - 1] = -abs(nums[abs_num - 1])

    for i in range(length):
        if nums[i] > 0:
            return i + 1

    return length + 1


nums1 = [1, 2, 0]
nums2 = [3, 4, -1, 1]
nums3 = [7, 8, 9, 11, 12]

print(first_missing_positive2(nums1))
print(first_missing_positive2(nums2))
print(first_missing_positive2(nums3))
